src/frontend/Grammar.mly

```diff
@@ -71,45 +95,87 @@ debug_filter:
       | "constraints" -> `Constraints
       | _ -> failwith "Invalid debug filter: try 'all' or 'constraints' " }
 
-atomic_econ:
+eproj:
+  | DOT FST
+    { E.Car }
+  | DOT SND
+    { E.Cdr }
+  | DOT; n = NUMERAL
+    { match n with
+      | 0 -> E.Car
+      | 1 -> E.Cdr
+      | _ -> failwith "Parser: invalid projection" }
+
+atom_econ:
+  | a = ATOM
+    { atom_to_econ a }
+  | a = ATOM; CARET; k = NUMERAL
+    { E.Var (a, k) }
+
+atomoid_econ:
   | BACKTICK; t = tm
     { E.Quo t }
   | a = HOLE_NAME;
     { E.Hole a }
-  | HOLE_NAME; LBR; e = eterm; RBR
+  | HOLE_NAME; LBR; e = located(econ); RBR
     { E.Guess e }
   | spec = univ_spec
     { let k, l = spec in E.Type (k, l) }
-  | LGL; es = separated_list(COMMA, eterm); RGL
+  (* in theory this rule can replace the following three, but it seems there's some bug.
+  | LPR; es = separated_list(COMMA, located(econ)); RPR
     { E.Tuple es }
-  | LPR; e = eterm; RPR
-    { e.con }
-  | a = ATOM; CARET; k = NUMERAL
-    { E.Var (a, k) }
-  | a = ATOM;
-    { if a = "_" then E.Hope else E.Var (a, 0) }
+  *)
+  | LPR; RPR
+    { E.Tuple [] }
+  | LPR; e = econ; RPR
+    { e }
+  | LPR; e = located(econ); COMMA; es = separated_nonempty_list(COMMA, located(econ)); RPR
+    { E.Tuple (e :: es) }
   | REFL
     { E.Refl }
   | n = NUMERAL;
     { E.Num n }
 
-atomic_eterm:
-  | e = atomic_econ
-    { eterm $startpos $endpos e }
+atomic:
+  | e = atom_econ
+    { e }
+  | e = atomoid_econ
+    { e }
 
-eframe:
-  | e = atomic_eterm
+framoid:
+  | e = located(atomoid_econ)
     { E.App e }
-  | DOT FST
-    { E.Car }
-  | DOT SND
-    { E.Cdr }
-  | DOT; n = NUMERAL
-    { match n with
-      | 0 -> E.Car
-      | 1 -> E.Cdr
-      | _ -> failwith "Parser: invalid projection" }
+  | p = eproj
+    { p }
 
+framic:
+  | e = located(atomic)
+    { E.App e }
+  | p = eproj
+    { p }
+
+spine:
```

Comment:
(BTW in case it wasn't clear, I'm not objecting to it! Just wanted to understand it)

Reply:
It's trying to keep atoms as `nonempty_list(ATOM)` till the last symbol. The critical case is the `d` position in the following state:
```
(a b c d : e) -> f
(a b c d .0)
(a b c d ^1)
...
```
Menhir refuses to make any decision before the symbol following `d`. Therefore, the `App` frames can only be reconstructed after we are sure `a b c d` are spines, not bound variables.

Reply:
The current version has the disadvantage that the individual location information of `b c d` in the above example is lost, because `nonempty_list(ATOM)` is not keeping any of it. The solution is to replace `nonempty_list(ATOM)` with something that remembers the location. @jonsterling Do you think it's important?

Reply:
@favonia OK, cool! Thanks very much for figuring this out. It's exactly the case that prevented me from being able to make the grammar how I wanted it using just my Weak Thought :laughing: 

For now, I'm OK with disrupting those source locations for now, we can clean it up in another pass later on. I'd be happy to do so.